samza-core/src/main/scala/org/apache/samza/job/local/ProcessJob.scala

```diff
@@ -19,113 +19,112 @@
-    ProcessJob.this
-  }
+  val lock = new Object
+  val processBuilder: ProcessBuilder = createProcessBuilder(commandBuilder)
+  var jobStatus: ApplicationStatus = New
+  var processThread: Option[Thread] = None
 
-  def waitForFinish(timeoutMs: Long) = {
-    val thread = new Thread {
-      setDaemon(true)
+
+  def submit: StreamJob = {
+    val threadStartCountDownLatch = new CountDownLatch(1)
+
+    // Create a non-daemon thread to make job runner block until the job finishes.
+    // Without this, the proc dies when job runner ends.
+    processThread = Some(new Thread {
       override def run {
+        var processExitCode = -1
+        var process: Option[Process] = None
+
+        setStatus(Running)
+
         try {
-          process.waitFor
+          threadStartCountDownLatch.countDown
+          process = Some(processBuilder.start)
+          processExitCode = process.get.waitFor
         } catch {
-          case e: InterruptedException => info("Got interrupt.", e)
+          case _: InterruptedException => process foreach { p => p.destroyForcibly }
+          case e: Exception => error("Encountered an error during job start: %s".format(e.getMessage))
+        } finally {
+          jobModelManager.stop
+          setStatus(if (processExitCode == 0) SuccessfulFinish else UnsuccessfulFinish)
         }
       }
-    }
+    })
 
-    thread.start
-    thread.join(timeoutMs)
-    thread.interrupt
-    jobStatus.getOrElse(null)
+    processThread.get.start
+    threadStartCountDownLatch.await
+    ProcessJob.this
   }
 
-  def waitForStatus(status: ApplicationStatus, timeoutMs: Long) = {
-    val start = System.currentTimeMillis
-
-    while (System.currentTimeMillis - start < timeoutMs && status != jobStatus) {
-      Thread.sleep(500)
+  def kill: StreamJob = {
+    if (getStatus == Running) {
+      processThread foreach { thread =>
+        thread.interrupt
+        thread.join
+      }
     }
+    ProcessJob.this
+  }
+
+  def waitForFinish(timeoutMs: Long): ApplicationStatus = {
+    require(timeoutMs >= 0, "Timeout values must be non-negative.")
 
-    jobStatus.getOrElse(null)
+    processThread foreach { thread => thread.join(timeoutMs) }
+    getStatus
   }
 
-  def getStatus = jobStatus.getOrElse(null)
-}
+  def waitForStatus(status: ApplicationStatus, timeoutMs: Long): ApplicationStatus = lock.synchronized {
+    require(timeoutMs >= 0, "Timeout values must be non-negative.")
 
-/**
- * Silly class to forward bytes from one stream to another. Using this to pipe
- * output from subprocess to this process' stdout/stderr.
- */
-class Piper(in: InputStream, out: OutputStream) extends Runnable {
-  def run() {
-    try {
-      val b = new Array[Byte](512)
-      var read = 1;
-      while (read > -1) {
-        read = in.read(b, 0, b.length)
-        if (read > -1) {
-          out.write(b, 0, read)
-          out.flush()
+    timeoutMs match {
+      case 0 => while (getStatus != status) lock.wait(0)
```

Comment:
Yes, it could block indefinitely. However:

1. It is no more hazardous than specifying a relatively large value, especially this is a long.
2. Using 0 to indicate waiting indefinitely would be consistent with other JDK threading/synchronization methods, e.g. `thread.wait()`, `thread.join()`, `object.wait()`.
3. In case someone would actually like to wait indefinitely, using 0 would probably be better than using, say, `Long.MaxValue`.

I will add logging right away.